### ISSUE TITLE
\ILLEGAL ARG in Tedit

### ISSUE THREAD

**Issue by masinter** [Interlisp/medley#1641]
to reproduce:
```
TEDIT(HELPSYS)
meta-A  (to select the whole content)
meta-C  (to copy the whole thing)
```
![image](https://github.com/Interlisp/medley/assets/1116587/eacd17c1-780b-4ebc-acad-55ac7efe9a8a)

in loadups from master (with or without merge-open-prs)

**Comment by nbriggs** [comment]
I see you mean "HELPSYS.TEDIT".  If you TEDIT the HELPSYS source file and do the same thing it breaks for a different reason -- you can't make a single string that is 88000 bytes long.

In your case, It's unclear how it should attempt to put a document with image objects into the system paste buffer when it has no idea what the destination is going be - inside medley or outside.

**Comment by rmkaplan** [comment]
In LFG we developed the code for what we called a “looked string”, basically a Tedit image object.  It encodes the contents of a Tedit selection including not just the characters but also the looks information and the image objects, with a display function that can render it via COPYINSERT in other destinations and a preprint function that shows the characters one by one.  It has no length limit.

I’ve been thinking about adding that to Tedit (there is still a stub (OBJECTOUTOFTEDIT) for the LFG function in the implementation of foreign copy (under the code for meta-C).  Given my new hard-won knowledge and new abstractions, it willl probably be much easier to add this capability:  the data comes essentially for free since a lot of simplification has come from new internal functions that isolate the pieces corresponding to a selection and then copy them outside of the B-tree.

But I will look at Larry’s report to see if something else is also broken.

**Comment by masinter** [comment]
no, i meant open a Interlisp EXEC and type `TEDIT(HELPSYS)` followed by Meta-A Meta-C. Starting from a new loadup of Master. 

![image](https://github.com/Interlisp/medley/assets/1116587/f7360189-7e49-4011-b216-0cf879b1a811)

**Comment by rmkaplan** [comment]
There has always (Venue sysout…) been an issue with the fact that the Tedit BIN function treats image objects as “characters” and returns them as it iterates down the document.  Callers (like TEDIT.SEL.AS.STRING) that expect to see real character codes fall over when they encounter these.

The GITFNS comparison uses COMPARETEXT, which also expects to see character codes, To work around that, I put in a hack in Tedit whereby you can specify in a property a charcode that you want BIN to return for every image object, instead of the image object itself (specified as 0 in the case of COMPARETEXT, to make it disappear).

Immediately, I can flush this error by having TEDIT.SEL.AS.STRING set the relevant property.

In the near term, it wouldn’t be too hard at least for TEDIT.SEL.AS.STRING to be a little more informative, by stuffing the result of the object’s preprint fn into the string.  And if it doesn’t have a preprintfn, maybe just a special sequence {OBJECT: xxx} where xxx is its getfn.  Or, again under a special property, the more general BIN could be made to detour into the characters of the preprintfn, and TEDIT.SEL.AS.STRING (and COMPARETEXT and other reading functions) would inherit that behavior.

In the long run, with a Tedit image object defined, the clipboard problem would resolve by not using TEDIT.SEL.AS.STRING but by providing a clipboard COPYINSERT function that does the “right” thing wrt image objects and looks for whatever properties the target clipboard may be able to deal with.

**Comment by rmkaplan** [comment]
One source of confusion may be that Tedit’s search strategy for the file to open is now driven of of *TEDIT-EXTENSIONS*, initially (TEDIT TED TXT TEXT BRAVO NIL).  So (TEDIT ‘HELPSY) would have a preference for HELPSYS.TEDIT if it can be found.  (TEDIT 'HELPSYS.) would force the Lisp source file (as would the command ts defined in TEDIT-PF-SEE).

So, Larry, are you pointing to the problem that you didn’t expect to see the documentation file, or the fact that meta-C breaks on a selection that is too long for a string or that includes image objects.

Another patch for the clipboard would be to replace TEDIT.SEL.AS.STRING in the copy function with a BIN loop that skips or does something special with image objects.  Maybe the Tedit clipboard functions should be moved into TEDIT?

**Comment by masinter** [comment]
I think you might be making this more complex. Currently, the Medley implementation of the copy-paste functionality is text only.  if you use the keyboard to copy things, you'll only get the text.

You could perhaps add a warning message to the prompt window if someone meta-c's a run including image objects, but otherwise just copy the text. But don't breakpoint. If you want to copy formatted text, use the click-destination & shift-click-source method of copying from one doc to another.

**Comment by rmkaplan** [comment]
Copy-paste (meta-C)  is one thing, we only get raw text unless the clipboard stream can do more (rtf?).  (I have a version of the paste function that at least includes a signature of an image object—could be made to issue a warning that it is doing that).  

But copying within Medley is separate, depends on what the receiving window can absorb (by its COPYINSERT function).

There is special code for copying from Tedit to Tedit, it thinks it can outsmart the COPYINSERT interface.  But other destination windows (OBJECTWINDOW, SKETCH…) can interpret all or most of the information provided with shift-select.  

It would be nice if meta-C could shadow the clipboard with its own internal full-formatted copy, so that meta-C would become a natural and complete way of moving things around even inside medley.

Meanwhile:  is there a limit to the number of characters or bytes that can go into the clipboard?  What happens if it is exceeded?  If there is a limit, can Medley know it, so that it can issue a warning?

**Comment by pamoroso** [comment]
I can't reproduce the issue in current Medley and the steps to select and copy the text work as expected.
